pandas/core/arrays/base.py

```diff
@@ -0,0 +1,194 @@
+        )
+
+    @abc.abstractmethod
+    def __iter__(self):
+        # type: () -> Iterator
+        pass
+
+    @abc.abstractmethod
+    def __len__(self):
+        # type: () -> int
+        pass
+
+    # ------------------------------------------------------------------------
+    # Required attributes
+    # ------------------------------------------------------------------------
+    @property
+    def base(self):
+        """The base array I am a view of. None by default."""
+
+    @property
+    @abc.abstractmethod
+    def dtype(self):
+        # type: () -> ExtensionDtype
+        """An instance of 'ExtensionDtype'."""
+
+    @property
+    def shape(self):
+        # type: () -> Tuple[int, ...]
+        return (len(self),)
+
+    @property
+    def ndim(self):
+        # type: () -> int
+        """Extension Arrays are only allowed to be 1-dimensional."""
+        return 1
+
+    @property
+    @abc.abstractmethod
+    def nbytes(self):
+        # type: () -> int
+        """The number of bytes needed to store this object in memory."""
+
+    # ------------------------------------------------------------------------
+    # Additional Methods
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def isna(self):
+        # type: () -> np.ndarray
+        """Boolean NumPy array indicating if each value is missing."""
+
+    # ------------------------------------------------------------------------
+    # Indexing methods
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def take(self, indexer, allow_fill=True, fill_value=None):
+        # type: (Sequence, bool, Optional[Any]) -> ExtensionArray
+        """For slicing"""
+
+    @abc.abstractmethod
+    def copy(self, deep=False):
+        # type: (bool) -> ExtensionArray
+        """Return a copy of the array."""
+
+    # ------------------------------------------------------------------------
+    # Block-related methods
+    # ------------------------------------------------------------------------
+    @property
+    def _fill_value(self):
+        # type: () -> Any
+        """The missing value for this type, e.g. np.nan"""
+        return None
+
+    @abc.abstractmethod
+    def _formatting_values(self):
+        # type: () -> np.ndarray
+        # At the moment, this has to be an array since we use result.dtype
+        """An array of values to be printed in, e.g. the Series repr"""
+
+    @classmethod
+    @abc.abstractmethod
+    def _concat_same_type(cls, to_concat):
+        # type: (Sequence[ExtensionArray]) -> ExtensionArray
+        """Concatenate multiple array
+
+        Parameters
+        ----------
+        to_concat : sequence of this type
+
+        Returns
+        -------
+        ExtensionArray
+        """
+
+    @abc.abstractmethod
+    def get_values(self):
+        # type: () -> np.ndarray
+        """A NumPy array representing your data.
```

Comment:
Let's clarify:
- This is used for `Series.values`, but otherwise is not used internally by pandas (This is true, right?).
- If I've implemented a custom scalar type, should this be an object array of containing all elements of that type? Or should I prefer something possibly lossy but faster? e.g., for IntervalArray, should I return an object array of Interval objects, or a structured array with two elements? It's OK if both options are fine depending on the circumstances, let's just give a little guidance.

Reply:
This is *not* used by `Series.values`, that currently returns the array, like `pd.Series(categorical)` is a `Categorical`. It seems to be used in

1. indexing, via `_values_from_object`, which seems undesirable
2. nanops, via `_values_from_object`, again undesirable

So let's hold off on discussing this until I can understand it better. We could potentially remove it. And if extension arrays are iterable, then we could just as well use `np.array(extension_array, dtype='object')` whenever we absolutely need an ndarray of objects.

Reply:
It is also used for eg concatting that ends up in object dtype, eg in case of concatting different types (a rather corner case). 
And it is also similar to `Series.get_values`.

I *think* in general this is the "fallback numpy array" in case we need one and we need to be able to infer what to do with it (so typically the object array of custom scalars).

It could be an option to use this to enable (or enable good error message) operations. For example that those values are used when the user calls for the `sum`. And if you then have an object array of custom scalars, this will be handled gracefully. 
Although (thinking while I am writing), this is in general also not ideal as we would want: a) the ability to let pandas operations dispatch to efficient implementations of the array itself (eg in hypothetical case you have a `sum` function on your custom array that does not need to construct object array of scalars) and b) the ability to disable numerical operations without the need to crystallize a full object array (as eg in geopandas this is very costly)

Reply:
> I think in general this is the "fallback numpy array" in case we need one and we need to be able to infer what to do with it (so typically the object array of custom scalars).

Yep. That sounds right.

> let pandas operations dispatch to efficient implementations of the array itself [...] and b) the ability to disable numerical operations without the need to crystallize a full object array

Agreed on both accounts, though I think getting reductions, cumulative ops, and numeric ops to all work properly on extension arrays will require some careful thought an quite a bit of work. At the moment, I'm hoping to just avoid *breaking* things like `df.mean()` when you have a mix of numeric and extension columns.

Reply:
See https://github.com/pandas-dev/pandas/issues/19318 for notes about `get_values` being called during indexing, specifically `Series.__getitem__`. I think it can be avoided.